### PR TITLE
Tachiyomi unit tests and fixes

### PR DESCRIPTION
**Pull request by ThePromidius**
# Fixed
- Fixed: Added a missing a check to return volume number instead of chapter number on single-file volumes for Tachiyomi APIs

### CODE REVIEW

File: API.Tests/Services/ReaderServiceTests.cs [hunk 1]
```diff
@@ -2226,4 +2226,133 @@ public void FormatChapterName_Comic_WithHash()
...
+        // Validate volumes chapter 0 have read status
+        //Assert.Equal(1, (await _unitOfWork.AppUserProgressRepository.GetUserProgressAsync(0, 1)).PagesRead);
+        Assert.Equal(1, (await _unitOfWork.AppUserProgressRepository.GetUserProgressAsync(5, 1)).PagesRead);
+        Assert.Equal(1, (await _unitOfWork.AppUserProgressRepository.GetUserProgressAsync(6, 1)).PagesRead);
+        // Validate next volume (2003) is not read
+        Assert.Null(await _unitOfWork.AppUserProgressRepository.GetUserProgressAsync(7, 1));
+
+    }
+    [Fact]
+    public async Task MarkVolumesUntilAsRead_ShouldMarkVolumesAsRead_2()
```
Comment (majora2007) [inline API.Tests/Services/ReaderServiceTests.cs:2295]: Try not to use _2 and instead give context

File: API.Tests/Services/ReaderServiceTests.cs [hunk 2]
```diff
@@ -2226,4 +2226,133 @@ public void FormatChapterName_Comic_WithHash()
...
+        await readerService.MarkVolumesUntilAsRead(user, 1, 2002);
+        await _context.SaveChangesAsync();
+
+        // Validate loose leaf chapters don't get marked as read
+        Assert.Null((await _unitOfWork.AppUserProgressRepository.GetUserProgressAsync(1, 1)));
+        Assert.Null((await _unitOfWork.AppUserProgressRepository.GetUserProgressAsync(2, 1)));
+        Assert.Null((await _unitOfWork.AppUserProgressRepository.GetUserProgressAsync(3, 1)));
+
+        // Validate volumes chapter 0 have read status
+        //Assert.Equal(1, (await _unitOfWork.AppUserProgressRepository.GetUserProgressAsync(0, 1)).PagesRead);
```
Comment (majora2007) [inline API.Tests/Services/ReaderServiceTests.cs:2287]: Remove dead code

File: API.Tests/Services/ReaderServiceTests.cs [hunk 3]
```diff
@@ -2226,4 +2226,133 @@ public void FormatChapterName_Comic_WithHash()
...
```
Comment (majora2007) [inline API.Tests/Services/ReaderServiceTests.cs:2350]: remove dead code

File: API/Controllers/TachiyomiController.cs [hunk 4]
```diff
@@ -39,53 +32,9 @@ public TachiyomiController(IUnitOfWork unitOfWork, IReaderService readerService,
     [HttpGet("latest-chapter")]
     public async Task<ActionResult<ChapterDto>> GetLatestChapter(int seriesId)
     {
+        if (seriesId < 1) return BadRequest("SeriesId must be greater than 1");
```
Comment (majora2007) [inline API/Controllers/TachiyomiController.cs:35]: A seriesId must be greater than 0.

File: API/Services/TachiyomiService.cs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+    public TachiyomiService(IUnitOfWork unitOfWork,IMapper mapper,IReaderService readerService)
+    {
+        _unitOfWork = unitOfWork;
+        _readerService = readerService;
+        _mapper = mapper;
+    }
+
+    public async Task<ChapterDto> GetLatestChapter(int seriesId,int userId)
+//    public async Task<ChapterDto> GetLatestChapter(int seriesId, int userId)
```
Comment (majora2007) [inline API/Services/TachiyomiService.cs:36]: Remove dead code

File: API/Services/TachiyomiService.cs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    private readonly CultureInfo _englishCulture = CultureInfo.CreateSpecificCulture("en-US");
+
+    public TachiyomiService(IUnitOfWork unitOfWork,IMapper mapper,IReaderService readerService)
+    {
+        _unitOfWork = unitOfWork;
+        _readerService = readerService;
+        _mapper = mapper;
+    }
+
+    public async Task<ChapterDto> GetLatestChapter(int seriesId,int userId)
```
Comment (majora2007) [inline API/Services/TachiyomiService.cs:35]: add a space after , (this goes for whole PR)

File: API/Services/TachiyomiService.cs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    Task<bool> MarkChaptersUntilAsRead(AppUser userWithProgress, int seriesId, float chapterNumber);
+}
+
+
+public class TachiyomiService : ITachiyomiService
+{
+    private readonly IUnitOfWork _unitOfWork;
+    private readonly IReaderService _readerService;
+    private readonly IMapper _mapper;
+    private readonly CultureInfo _englishCulture = CultureInfo.CreateSpecificCulture("en-US");
```
Comment (majora2007) [inline API/Services/TachiyomiService.cs:26]: Make this static

File: API/Services/TachiyomiService.cs (hunk 6 above)
Comment (majora2007) [inline API/Services/TachiyomiService.cs:35]: Add Documentation for the service (you should copy from controller)

File: API/Services/TachiyomiService.cs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+            var looseLeafChapterVolume = volumes.FirstOrDefault(v => v.Number == 0);
+            if (looseLeafChapterVolume == null)
+            {
+                var volumeChapter = _mapper.Map<ChapterDto>(volumes.Last().Chapters.OrderBy(c => float.Parse(c.Number), ChapterSortComparerZeroFirst.Default).Last());
+                if (volumeChapter.Number == "0")
+                {
+                    var volume = volumes.First(v => v.Id == volumeChapter.VolumeId);
+                    return new ChapterDto()
+                    {
+                        Number = (volume.Number / 10000f).ToString("R", _englishCulture)
```
Comment (majora2007) [inline API/Services/TachiyomiService.cs:65]: Write a comment about what R means and link to the MS doc

File: API/Services/TachiyomiService.cs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+                    };
+                }
+
+                return new ChapterDto()
+                {
+                    Number = (int.Parse(volumeChapter.Number) / 10000f).ToString("R", _englishCulture)
+                    //Number = $"{int.Parse(volumeChapter.Number) / 1000f}"
```
Comment (majora2007) [inline API/Services/TachiyomiService.cs:72]: Remove dead code

File: API/Services/TachiyomiService.cs [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+                    var volume = volumes.First(v => v.Id == volumeChapter.VolumeId);
+                    return new ChapterDto()
+                    {
+                        Number = (volume.Number / 10000f).ToString("R", _englishCulture)
+                    };
+                }
+
+                return new ChapterDto()
+                {
+                    Number = (int.Parse(volumeChapter.Number) / 10000f).ToString("R", _englishCulture)
```
Comment (majora2007) [inline API/Services/TachiyomiService.cs:71]: Add a <remarks> in the method documentation to explain how we handle the / 10000. 

Also change 10000f -> 10_000f

File: API/Services/TachiyomiService.cs [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        switch (chapterNumber)
+        {
+            // When Tachiyomi sync's progress, if there is no current progress in Tachiyomi, 0.0f is sent.
+            // Due to the encoding for volumes, this marks all chapters in volume 0 (loose chapters) as read.
+            // Hence we catch and return early, so we ignore the request.
+            case 0.0f:
+                return true;
+            case < 1.0f:
+            {
+                // This is a hack to track volume number. We need to map it back by x100
```
Comment (majora2007) [inline API/Services/TachiyomiService.cs:113]: Comment no longer matches implementation.

File: API/Services/TachiyomiService.cs [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+                var chapterString = $"{chapterNumber}";
+                var volumeNumber = int.Parse($"{(int)(chapterNumber * 10000)}", _englishCulture);
+                //var volumeNumber = int.Parse($"{float.Parse(chapterString.Substring(0, Math.Min(7, chapterString.Length))) * 100f}");
```
Comment (majora2007) [inline API/Services/TachiyomiService.cs:116]: Remove dead code

File: API/Services/TachiyomiService.cs [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+                Number = (volumeWithProgress.Number / 10000f).ToString("R", _englishCulture)
+
+            };
+        }
+
+        // Progress is just on a chapter, return as is
+        return prevChapter;
+    }
+
+    public async Task<bool> MarkChaptersUntilAsRead(AppUser userWithProgress, int seriesId, float chapterNumber)
```
Comment (majora2007) [inline API/Services/TachiyomiService.cs:100]: In the method documentation, you need to mention this saves to the DB

**Review by majora2007 — CHANGES_REQUESTED** [review]

File: API/Controllers/TachiyomiController.cs [hunk 14]
```diff
@@ -21,14 +16,12 @@ namespace API.Controllers;
 public class TachiyomiController : BaseApiController
 {
     private readonly IUnitOfWork _unitOfWork;
-    private readonly IReaderService _readerService;
-    private readonly IMapper _mapper;
+    private readonly ITachiyomiService _tachiyomiService;
 
-    public TachiyomiController(IUnitOfWork unitOfWork, IReaderService readerService, IMapper mapper)
+    public TachiyomiController(IUnitOfWork unitOfWork, ITachiyomiService tachiyomiService, IMapper mapper)
```
Comment (majora2007) [inline API/Controllers/TachiyomiController.cs:21]: I still see Mapper injected, but it's not needed anymore

File: API/Controllers/TachiyomiController.cs (hunk 4 above)
Comment (majora2007) [inline API/Controllers/TachiyomiController.cs:35]: return BadRequest("SeriesId must be greater than 0");

File: API/Services/TachiyomiService.cs [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+                break;
+        }
+
+
+        _unitOfWork.UserRepository.Update(userWithProgress);
+
+        if (!_unitOfWork.HasChanges()) return true;
+        if (await _unitOfWork.CommitAsync()) return true;
+
+        await _unitOfWork.RollbackAsync();
```
Comment (majora2007) [inline API/Services/TachiyomiService.cs:131]: We need a try catch so you can catch the exception and log it out on Error(ex, "There was an issue updating...")

**Review by majora2007 — CHANGES_REQUESTED** [review]
There are still pending comments

File: API/Controllers/TachiyomiController.cs (hunk 14 above)
Comment (ThePromidius) [reply API/Controllers/TachiyomiController.cs:21]: That's weird. I don't see it on mine

**Review by majora2007 — APPROVED** [review]

**Pull request merged**